File: tree/leetcode_Binary_Tree_Zigzag_Level_Order_Traversal.java
/*Binary_Tree_Zigzag_Level_Order_Travesral

Given a binary tree, return the zigzag level order traversal of its nodes'
values. (ie, from left to right, then right to left for the next level and
alternate between).
For example:
Given binary tree {3,9,20,#,#,15,7},
return its zigzag level order traversal as:
[
    [3],
    [20,9],
    [15,7]
]
*/

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;


public class leetcode_Binary_Tree_Zigzag_Level_Order_Travesral {

    public static void main(String[] args) {

    }

    public static ArrayList<ArrayList<Integer>> zigzagLevelOrder(TreeNode root) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if (root == null) {
			return result;
		}
        Queue<TreeNode> unvisited = new LinkedList<TreeNode>();
        ArrayList<Integer> tmp = new ArrayList<Integer>();
        unvisited.add(root);
        int count = 1, reverse = 0, swap, head, tail;
        while (!unvisited.isEmpty()) {
            TreeNode node = unvisited.remove();
            tmp.add(node.val);
            if (node.left != null) {
				unvisited.add(node.left);
			}
            if (node.right != null) {
				unvisited.add(node.right);
			}
            if (--count == 0) {
                if (reverse++ % 2 == 1) {
                    head = 0;
                    tail = tmp.size() - 1;
                    while (head < tail) {
                        swap = tmp.get(head);
                        tmp.set(head++, tmp.get(tail));
                        tmp.set(tail--, swap);
                    }
                }
                result.add(tmp);
                tmp = new ArrayList<Integer>();
                count = unvisited.size();
            }
        }
        return result;        
    }
}


class TreeNode {
    TreeNode left;
    TreeNode right;
    int val;

    TreeNode(int val) {
		this.left = null;
		this.right = null;
        this.val = x;
    }
}

File: tree/leetcode_Binary_Tree_Zigzag_Level_Order_Traversal.cpp
/**
 * Copyright 2016
 * author: madarfacar
 *
 * binary tree zigzag level order traversal
 */

#include <stdio.h>
#include <vector>

using std::vector;


/**
 * Definition for a binary tree node.
 */
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
 public:
  vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
    vector<vector<int>> levels;
    if (root == NULL) {
      return levels;
    }

    vector<TreeNode*> q {root};
    vector<TreeNode*> n;
    vector<int> level;
    bool zigzag = false;

    while (!q.empty()) {
      TreeNode* node = q.back();
      q.pop_back();
      level.push_back(node->val);
      if (!zigzag) {
        if (node->left != NULL) {
          n.push_back(node->left);
        }
        if (node->right != NULL) {
          n.push_back(node->right);
        }
      } else {
        if (node->right != NULL) {
          n.push_back(node->right);
        }
        if (node->left != NULL) {
          n.push_back(node->left);
        }
      }

      if (q.empty()) {
        if (!level.empty()) {
          levels.push_back(level);
          level = vector<int> ();
        }
        q = n;
        n = vector<TreeNode*> ();
        zigzag = !zigzag;
      }
    }
    return levels;
  }
};


int main() {
  Solution sol;

  TreeNode* head = new TreeNode(1);
  head->left = new TreeNode(2);
  head->right = new TreeNode(3);
  head->left->left = new TreeNode(4);
  head->left->right = new TreeNode(5);
  head->right->left = new TreeNode(6);
  head->right->right = new TreeNode(7);

  vector<vector<int>> levels = sol.zigzagLevelOrder(head);

  for (const vector<int> level : levels) {
    for (int val : level) {
      printf("%d ", val);
    }
    printf("\n");
  }
}
